### ISSUE TITLE
Widen JustFloat to `float | int` as part of widening literal types

### ISSUE THREAD

**Issue by Wizzerinus** [astral-sh/ty#2226]
### Summary

In the following example, the type for `FloatDict` is inferred as `dict[int, float]` rather than `dict[int, float | int]`, which is a bit unexpected to me, because that causes the assignment to fail typechecking.

```py
FloatDict = dict.fromkeys(range(10), 0.5)

def make_float() -> float:
    raise NotImplementedError()

FloatDict[0] = make_float()
```

Diagnostic:
```
error[invalid-assignment]: Invalid subscript assignment with key of type `Literal[0]` and value of type `int | float` on object of type `dict[int, float]`
 --> floatdict.py:6:1
  |
4 |     raise NotImplementedError()
5 |
6 | FloatDict[0] = make_float()
  | ^^^^^^^^^^^^^^^------------
  |                |
  |                Expected value of type `float`, got `int | float`
  |
info: rule `invalid-assignment` is enabled by default

Found 1 diagnostic
```

### Version

ty 0.0.7

**Comment by carljm** [comment]
Thanks! I think we need to always "widen" float inferences to `float | int` in the same scenarios where we widen literal types. So just like we already infer `dict.fromkeys(range(10), 1)` as `dict[int, int]` (not `dict[int, Literal[1]]`), we should likewise infer this case as `dict[int, float | int]` not `dict[int, float]`.

**Comment by carljm** [comment]
I think this should be a fairly easy change -- we already have the literal-widening logic, this _should_ be just adding one more case to it.
